docs/DevGuide/TensorExpressions.md

```diff
@@ -0,0 +1,411 @@
+\cond NEVER
+Distributed under the MIT License.
+See LICENSE.txt for details.
+\endcond
+# Writing tensor equations with TensorExpressions {#writing_tensorexpressions}
+
+\tableofcontents
+
+SpECTRE's `TensorExpression`s interface allows you to write tensor equations in
+SpECTRE in C++ with syntax that resembles tensor index notation. To use it,
+simply add this include to the top of your file:
+```
+#include "DataStructures/Tensor/Tensor.hpp"
+```
+The following guide assumes a basic understanding of the `Tensor` class and
+\ref tnsr "tnsr" type aliases.
+
+# Syntax {#te_syntax}
+`TensorExpression`s are arithmetic expressions of `Tensor`s that can be
+evaluated using `tenex::evaluate`. Terms used in the expression may be `Tensor`s
+or numbers (see [supported types](#te_data_type_support)).
+
+As a simple example of how `TensorExpression`s are used, if you would like to
+raise the index of some `Tensor` `R` with some spacetime metric `Tensor` `g`,
+i.e. \f$R^c{}_b = R_{ab} g^{ac}\f$, you can compute this with
+`TensorExpression`s by doing:
+```
+auto R_up = tenex::evaluate<ti::C, ti::b>(R(ti::a, ti::b) * g(ti::A, ti::C));
+```
+
+where `R_up`, `R`, and `g` are rank 2 spacetime `Tensor`s and the `ti::*`
+variables are `TensorIndex`s. Here, the argument to
+\ref tenex::evaluate "evaluate" is the RHS tensor expression to compute, `R_up`
+is the result LHS `Tensor`, and the template arguments to
+\ref tenex::evaluate "evaluate" are the LHS `Tensor`'s indices. The LHS symmetry
+will be deduced from the RHS tensors' symmetries and order of operations.
+
+Alternatively, if you already have a LHS `Tensor` variable, you can pass it into
+the following \ref tenex::evaluate "evaluate" overload, where the LHS `Tensor`
+provided will be assigned to the result of the RHS expression:
+
+```
+tenex::evaluate<ti::C, ti::b>(
+    make_not_null(&R_up), R(ti::a, ti::b) * g(ti::A, ti::C));
+```
+
+One advantage of this overload is that it uses the symmetry and index types
+(spatial/spacetime) of the provided LHS tensor instead of deducing them from the
+RHS expression, so this enables you to specify the LHS index structure in cases
```

Comment:
Why do you still need the `ti::C, ti::b` template parameters if the symmetries and index types of the LHS tensor are used?

Reply:
To specify the order of the LHS indices. You may want `L^c{}_b` or you may want `L_b{}^c`. If you think a note should be added, let me know

Reply:
To specify the order of the LHS indices. You may want `L^c{}_b` or you may want `L_b{}^c`. In this case, the index types of `b` and `c` are different, so yes at compile time, the order could be assigned for you, but this isn't the case if you have LHS indices of the same type. For example:

```
tnsr::ab<...> L{};
tnsr::ab<...> R{};
tnsr::ab<...> S{};

// this wouldn't compile, just what it would look like without the LHS index tparams
tenex::evaluate(make_not_null(&L), R(ti::a, ti::b) + S(ti::a, ti::b));
```
You might want the LHS to be `L_{ab}` or maybe you want it `L_ba`, so the `evaluate` tparams let you specify that order. And while the order could be inferred in cases where the index types are different, that handling isn't something I've implemented and I think the way it is now also means there is a unified interface / one way you do things and forces people to be explicit

Reply:
In this case, the index types of `b` and `c` are different, so yes at compile time, the order could be deduced for you in theory, but this isn't the case if you have LHS indices of the same type. For example:

```
tnsr::ab<...> L{};
tnsr::ab<...> R{};
tnsr::ab<...> S{};

// this wouldn't compile right now, this is just what it would look like without the LHS index tparams
tenex::evaluate(make_not_null(&L), R(ti::a, ti::b) + S(ti::a, ti::b));
```
You might want the LHS to be `L_{ab}` or maybe you want it `L_ba`, so the `evaluate` tparams let you specify that order. And while the order could be inferred in cases where the index types are different (like the raising index example), doing that deduction isn't something I've implemented and I think the way it is now also means there is a unified interface / one way you do things and forces people to be explicit

So yes, the LHS index order of this example could be deduced, but I don't think we want to

Reply:
Sounds good! Thanks for explaining! Could you add a note here then saying something like

```
Even though we use the LHS tensors index types rather than deducing them from the
RHS, for technical reasons, you must always specify the tparams to `evaluate` for non-Scalar
tensors.
```